src/PrestaShopBundle/Translation/Provider/ExternalModuleLegacySystemProvider.php

```diff
@@ -225,10 +245,82 @@ public function getMessageCatalogue()
     }
 
     /**
-     * @return string returns the module domain
+     * Replaces dots in the catalogue's domain names
+     * and filters out domains not corresponding to the one from this module
+     *
+     * @param MessageCatalogueInterface $catalogue
+     *
+     * @return MessageCatalogue
+     */
+    private function filterDomains(MessageCatalogueInterface $catalogue) {
+        $normalizer = new DomainNormalizer();
+        $newCatalogue = new MessageCatalogue($catalogue->getLocale());
+
+        // add delimiter to
+        $validTranslationDomains = $this->getFilters();
+
+        foreach ($catalogue->getDomains() as $domain) {
+            // remove dots
+            $newDomain = $normalizer->normalize($domain);
+
+            // only add if the domain is relevant to this module
+            foreach ($validTranslationDomains as $pattern) {
+                if (preg_match($pattern, $newDomain)) {
+                    $newCatalogue->add(
+                        $catalogue->all($domain),
+                        $newDomain
+                    );
+                    break;
+                }
+            }
+        }
+
+        return $newCatalogue;
+    }
+
+    /**
+     * Builds the default catalogue
+     *
+     * @return MessageCatalogue
+     */
+    private function buildFreshDefaultCatalogue()
+    {
+        $defaultCatalogue = new MessageCatalogue($this->locale);
+
+        try {
+            // look up files in the core translations
+            $defaultCatalogue = $this->moduleProvider
+                ->setModuleName($this->moduleName)
+                ->setLocale($this->locale)
+                ->getDefaultCatalogue();
+        } catch (FileNotFoundException $exception) {
+            // there are no xliff files for this module in the core
+        }
+
+        try {
+            // analyze files and extract wordings
+            $additionalDefaultCatalogue = $this->legacyModuleExtractor->extract($this->moduleName, $this->locale);
+            $defaultCatalogue = $this->filterDomains($additionalDefaultCatalogue);
+        } catch (UnsupportedLocaleException $exception) {
+            // Do nothing as support of legacy files is deprecated
+        }
+
+        return $defaultCatalogue;
+    }
+
+    /**
+     * Returns the cached default catalogue
+     *
+     * @return MessageCatalogue
      */
-    private function getModuleDomain()
+    private function getCachedDefaultCatalogue()
```

Comment:
are we reinventing cache system here when I wanted to work on a LegacyFileProvider compatible with the Symfony system which provide a cache system by default? :trollface:

Reply:
I don't understand what you're trying to say. There was no cache before, this one exists only to avoid extracting the catalogue from source code twice per page load (once in `getDefaultCatalogue()`, once in `getXliffCatalogue()`).

FYI this extraction is performed EVERY TIME a bit of information is loaded. That is:

- Once to display the tree
- Once to display the actual translations
- Once every time you click on a tree leaf

This should be improved by storing this information in a file cache, which in itself it's simple. But then again there's the problem of cache invalidation. Ideally, catalog extraction (that is, source code analaysis) should be performed once, then triggered manually via a button. An alternative would be to invalidate the cache every time the page is loaded...

Anyway, once the cache situation is resolved, then the catalogue could be interpreted once, stored in cache, and work as just another symfony catalogue.

Reply:
This could probably be done using `ConfigCache` with a dump file, and providing the list of module files as resources. This way when a file is modified the cache would be automatically refreshed.

Reply:
Interesting idea, but...

> In debug mode, a .meta file will be created in the same directory as the cache file itself. This .meta file contains the serialized resources, whose timestamps are used to determine if the cache is still fresh. **When not in debug mode, the cache is considered to be "fresh" as soon as it exists, and therefore no .meta file will be generated.**

([source](https://symfony.com/doc/3.4/components/config/caching.html))

This means that cache invalidation must be performed manually, which makes implementation not-so-easy. 

Furthermore, the cache object requires a list of all the files it has to track for debug-mode invalidation, but the extractor doesn't know which files it must inspect, it just crawls files in a path, then it inspects them. It doesn't inform which files it inspected either.

Reply:
The cache system would have to crawl the filesystem every time to find out if there are new files it hasn't discovered. It can be done as it would save us from re extracting the catalogue, but the implementation would be particularly tricky as the lookup is performed by the extractor itself, and also... that class belongs to TranslationToolsBundle 😅

It would be much easier just to store a status flag that would be disabled manually via trigger (aka a button). Less magic, and arguably better UX.